contracts/governance/InterchainMultisig.sol

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import { IInterchainMultisig } from '../interfaces/IInterchainMultisig.sol';
+import { BaseWeightedMultisig } from './BaseWeightedMultisig.sol';
+import { SafeNativeTransfer } from '../libs/SafeNativeTransfer.sol';
+import { Caller } from '../utils/Caller.sol';
+
+/**
+ * @title Multisig Contract
+ * @notice An extension of MultisigBase that can call functions on any contract.
+ */
+contract InterchainMultisig is Caller, BaseWeightedMultisig, IInterchainMultisig {
+    // keccak256('InterchainMultisig.Storage')
+    bytes32 private constant INTERCHAIN_MULTISIG_INTERCHAINMULTISIGSTORAGE_LOCATION =
+        0x5a9dc2248a56f285e6221da431581a5990380ebfa07727563571b2be1517a23e;
+
+    using SafeNativeTransfer for address;
+
+    bytes32 public immutable chainNameHash;
+
+    struct InterchainMultisigStorage {
+        mapping(bytes32 => bool) isPayloadExecuted;
+    }
+
+    /**
+     * @notice Contract constructor
+     * @dev Sets the initial list of signers and corresponding threshold.
+     * @param chainName The name of the chain
+     * @param weightedSigners The weighted signers payload
+     */
+    constructor(string memory chainName, bytes memory weightedSigners) {
+        chainNameHash = keccak256(bytes(chainName));
+
+        _rotateSigners(weightedSigners);
+    }
+
+    modifier onlySelf() {
+        if (msg.sender != address(this)) revert NotSelf();
+
+        _;
+    }
+
+    /**
+     * @notice Checks if a payload has been executed
+     * @param payloadHash The hash of the payload payload
+     * @return True if the payload has been executed
+     */
+    function isPayloadExecuted(bytes32 payloadHash) external view returns (bool) {
+        return _interchainMultisigStorage().isPayloadExecuted[payloadHash];
+    }
+
+    /**
+     * @notice Executes an external contract call.
+     * @notice This function is protected by the onlySigners requirement.
+     * @dev Calls a target address with specified calldata and passing provided native value.
+     * @param batch The batch of calls to execute
+     * @param weightedSigners The weighted signers payload
+     * @param signatures The signatures payload
+     */
+    function executeCalls(
+        bytes calldata batch,
+        bytes calldata weightedSigners,
+        bytes[] calldata signatures
+    ) external payable {
+        bytes32 payloadHash = keccak256(batch);
+        bool isLatestSigners = _validateProof(payloadHash, weightedSigners, signatures);
```

Comment:
since we will set the retention to 1 (always pass 1 in the constructor to base contract), and only allow latest signers, the complicated rotation handling logic isn't needed. You can add an unit test to verify old signer can't execute.